### PR TITLE
eval: split TestTransactionGroup from BlockEvaluator using TransactionGroupTester

### PR DESCRIPTION
**Pull request by cce**
## Summary

eval.TestTransactionGroup (and TestTransaction) need only read-only access to a few details about the next pending block. This introduces an interface TestEvalContext that is used by the transaction pool, so that there can be parallel TX backlog workers checking if transactions are well-formed and move them from the backlogQueue to the postVerificationQueue.

## Test Plan

Updated TxHandler test that was checking Remember() vs Test() and counting metrics bumped by different kinds of err returns.

### CODE REVIEW

**Review by gmalouf — DISMISSED** [review]

**Review by jannotti — COMMENTED** [review]
Seems like a good idea, I had small comments, but willing to approve as is.

File: data/txHandler.go [hunk 1]
```diff
@@ -268,7 +268,8 @@ func (handler *TxHandler) Start() {
 		},
 	})
 
-	handler.backlogWg.Add(2)
+	handler.backlogWg.Add(3)
+	go handler.backlogWorker()
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/txHandler.go:272]: I don't see any useful thing to do here in the short term except pick a random small number like 2 or 3, so I vote for resolving.

File: data/pools/transactionPool.go [hunk 2]
```diff
@@ -114,6 +117,36 @@ type BlockEvaluator interface {
...
+// calls to TestTransactionGroup for candidate transactions calling transactionPool.Test.
+type testEvalContext struct {
+	ledger   *ledger.Ledger
+	block    bookkeeping.Block
+	proto    config.ConsensusParams
+	specials transactions.SpecialAddresses
+}
+
+func newTestBlockEvaluator(ledger *ledger.Ledger, block bookkeeping.Block) *eval.TestBlockEvaluator {
+	return &eval.TestBlockEvaluator{
```
[the remark this answers is not in the document]
Comment (jannotti) [reply data/pools/transactionPool.go:130]: Maybe we're coming to agreement, I can't tell. I get why you introduced the interface, so you could hand the smaller thing to `TestTransactionGroup`, but I was wondering about the extra wrapping in another struct, `TestBlockEvaluator`. Maybe what happened is that someone previously introduced `TestBlockEvaluator`, and you're just trying to make things work with the new interface. I think the `TestBlockEvaluator` can just go away, and you can promote the transaction pool's `testEvalContext` to an exported type. That type would implement your new interface, and that new interface would, itself, have `TestTransactionGroup` on it. I _think_ that's where you're going with your last few comments in your response. (Except you seem to be thinking about adding another type "SomethingNew", where I'd think you can make that a method on your new interface.

I find all of the names of all of these types to be very hard to follow, so I can't even tell if I'm making sense. There's no real information in the names that distinguish their uses.

[the remark this answers is not in the document]
Comment (jannotti) [reply data/pools/transactionPool.go:130]: Maybe I'm pushing toward an interface like `GroupTester`, with a single `TestTransactionGroup()` method.  Would that imply too much repetition of the existing `TestTransactionGroup()` function?

**Review by jannotti — APPROVED** [review]

**Review by gmalouf — APPROVED** [review]

File: ledger/eval/eval.go [hunk 3]
```diff
@@ -1163,6 +1188,7 @@ func (eval *BlockEvaluator) transaction(txn transactions.SignedTxn, evalParams *
 		}
 
 		// Transaction already in the ledger?
+		// this checks against the txns added to this evaluator; testTransaction currently only checks against committed txns.
```
Comment (algorandskiy) [inline ledger/eval/eval.go:1191]: > // this checks against the txns added to this evaluator; testTransaction currently only checks against committed txns.

This is not 100% accurate, `checkDup` checks the ledger as well, maybe reword to "against the txns added to this evaluator and the ledger" ?

**Review by algorandskiy — APPROVED** [review]
So the idea is to have `txpool.Test` executed without taking `pool.mu` so do not block actual evaluation. LGTM
